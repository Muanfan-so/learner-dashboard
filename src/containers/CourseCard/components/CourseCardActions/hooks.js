import { Locked } from '@edx/paragon/icons';
import { useIntl } from '@edx/frontend-platform/i18n';

import { hooks as appHooks } from 'data/redux';
import useSelectSession from 'containers/SelectSession/hooks';
import messages from './messages';

export const useCardActionData = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const {
    canUpgrade,
    isAudit,
    isAuditAccessExpired,
    isVerified,
  } = appHooks.useCardEnrollmentData(courseNumber);
  const { isPending, isArchived } = appHooks.useCardCourseRunData(courseNumber);
  const {
    isEntitlement,
    canViewCourse,
    isFulfilled,
    isExpired,
    canChange,
    hasSessions,
  } = appHooks.useCardEntitlementsData(courseNumber);
  const { openSessionModal } = useSelectSession({ courseNumber });

  let primary;
  let secondary = null;
  if (isEntitlement) {
    if (!isFulfilled) {
      primary = {
        children: formatMessage(messages.selectSession),
        disabled: !(canChange && hasSessions),
        onClick: openSessionModal,
      };
    } else {
      primary = {
        children: formatMessage(messages.viewCourse),
        disabled: !canViewCourse || isExpired,
      };
    }
  } else {
    if (!isVerified) {
      secondary = {
        iconBefore: Locked,
        variant: 'outline-primary',
        disabled: !canUpgrade,
        children: formatMessage(messages.upgrade),
      };
    }
    if (isPending) {
      primary = { children: formatMessage(messages.beginCourse) };
    } else if (!isArchived) {
      primary = {
        children: formatMessage(messages.resume),
        disabled: isAudit && isAuditAccessExpired,
      };
    } else {
      primary = { children: formatMessage(messages.viewCourse) };
    }
  }
  return { primary, secondary };
};

export default useCardActionData;
