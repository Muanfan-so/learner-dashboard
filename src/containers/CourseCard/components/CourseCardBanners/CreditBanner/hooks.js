import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = appHooks.useCardCreditData(cardId);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  if (!credit.isEligible) { return null; }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
