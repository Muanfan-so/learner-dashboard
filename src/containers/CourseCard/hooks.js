import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

export const useCardData = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const { title, bannerUrl } = appHooks.useCardCourseData(courseNumber);

  return {
    title,
    bannerUrl,
    formatMessage,
  };
};

export default useCardData;
