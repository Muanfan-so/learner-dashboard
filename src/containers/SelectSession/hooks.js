import { hooks as appHooks, actions } from 'data/redux';

import { useDispatch } from 'react-redux';

export const useSelectSession = ({ courseNumber }) => {
  const dispatch = useDispatch();
  const {
    showSessionModal,
    showLeaveSessionInSessionModal,
  } = appHooks.useSelectSessionsModalData();

  const { entitlementSessions } = appHooks.useCardEntitlementsData(courseNumber);

  const { title: courseTitle } = appHooks.useCardCourseData(courseNumber);

  const updateSessionModal = (showModal, showLeaveOption = false) => dispatch(
    actions.app.updateSelectSessionModal({
      showSessionModal: showModal,
      showLeaveSessionInSessionModal: showLeaveOption,
      courseNumber,
    }),
  );

  return {
    showSessionModal,
    closeSessionModal: () => updateSessionModal(false),
    openSessionModal: () => updateSessionModal(true),
    openSessionModalWithLeaveOption: () => updateSessionModal(true, true),
    showLeaveSessionInSessionModal,
    entitlementSessions,
    courseTitle,
  };
};

export default useSelectSession;
