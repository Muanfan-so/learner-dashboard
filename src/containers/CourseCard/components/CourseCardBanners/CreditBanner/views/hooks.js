import React from 'react';

import { StrictDict } from 'utils';
import { apiHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest().then(setRequestData);
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
