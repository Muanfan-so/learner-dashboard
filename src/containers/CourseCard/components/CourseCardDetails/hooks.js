import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';
import useSelectSession from 'containers/SelectSession/hooks';

import * as module from './hooks';
import messages from './messages';

export const useAccessMessage = ({ courseNumber }) => {
  const { formatMessage, formatDate } = useIntl();
  const {
    accessExpirationDate,
    isAudit,
    isAuditAccessExpired,
  } = appHooks.useCardEnrollmentData(courseNumber);
  const { isArchived, endDate } = appHooks.useCardCourseRunData(courseNumber);

  if (isAudit) {
    return formatMessage(
      isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
      { accessExpirationDate: formatDate(accessExpirationDate) },
    );
  }

  return formatMessage(
    isArchived ? messages.courseEnded : messages.courseEnds,
    { endDate: formatDate(endDate) },
  );
};

export const useCardDetailsData = ({ courseNumber }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(courseNumber).name;
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementsData(courseNumber);

  const { openSessionModalWithLeaveOption: openSessionModal } = useSelectSession({ courseNumber });

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: module.useAccessMessage({ courseNumber }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    formatMessage,
  };
};

export default useCardDetailsData;
