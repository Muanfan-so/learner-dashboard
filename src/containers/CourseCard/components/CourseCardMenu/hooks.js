import React from 'react';
import { StrictDict } from 'utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val), // eslint-disable-line
  isEmailSettingsVisible: (val) => React.useState(val), // eslint-disable-line
});

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const eventName = track.course.courseOptionsDropdownClicked;
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(eventName, cardId);
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useCourseCardMenu = (cardId) => {
  const { courseName } = reduxHooks.useCardCourseData(cardId);
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);
  const handleTwitterShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'twitter',
  );
  const handleFacebookShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'facebook',
  );

  const showUnenrollItem = isEnrolled && !isEarned;
  const showDropdown = showUnenrollItem || isEmailEnabled || facebook.isEnabled || twitter.isEnabled;

  return {
    courseName,
    isMasquerading,
    isEmailEnabled,
    showUnenrollItem,
    showDropdown,
    facebook,
    twitter,
    handleTwitterShare,
    handleFacebookShare,
  };
};
