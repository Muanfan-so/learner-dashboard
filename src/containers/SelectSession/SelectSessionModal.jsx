import React from 'react';
import PropTypes from 'prop-types';

import { useIntl } from '@edx/frontend-platform/i18n';
import {
  ActionRow, Button, Form, ModalDialog,
} from '@edx/paragon';

import { nullMethod } from 'hooks';
import { dateFormatter } from 'utils';

import useSelectSession from './hooks';
import messages from './messages';

export const SelectSessionModal = ({ courseNumber }) => {
  const {
    entitlementSessions,
    showSessionModal,
    closeSessionModal,
    showLeaveSessionInSessionModal,
    courseTitle,
  } = useSelectSession({
    courseNumber,
  });

  const { formatMessage, formatDate } = useIntl();

  let header;
  let hint;
  if (showLeaveSessionInSessionModal) {
    header = formatMessage(messages.changeOrLeaveHeader);
    hint = formatMessage(messages.changeOrLeaveHint);
  } else {
    header = formatMessage(messages.selectSessionHeader, {
      courseTitle,
    });
    hint = formatMessage(messages.selectSessionHint);
  }

  return (
    <ModalDialog
      isOpen={showSessionModal}
      onClose={nullMethod}
      hasCloseButton={false}
      title=""
    >
      <div
        className="bg-white p-3 rounded shadow"
        size="md"
        style={{ textAlign: 'start' }}
      >
        <h4>{header}</h4>
        <Form.Group>
          <Form.Label>{hint}</Form.Label>
          <Form.RadioSet name="sessions">
            {entitlementSessions?.map((entitle) => (
              <Form.Radio key={entitle.startDate} value={entitle.startDate}>
                {dateFormatter(formatDate, entitle.startDate)} - {dateFormatter(formatDate, entitle.endDate)}
              </Form.Radio>
            ))}
            {showLeaveSessionInSessionModal ? (
              <Form.Radio value="leave">
                {formatMessage(messages.leaveSessionOption)}
              </Form.Radio>
            ) : null}
          </Form.RadioSet>
        </Form.Group>
        <ActionRow>
          <Button variant="tertiary" onClick={closeSessionModal}>
            {formatMessage(messages.nevermind)}
          </Button>
          <Button>{formatMessage(messages.confirmSession)}</Button>
        </ActionRow>
      </div>
    </ModalDialog>
  );
};
SelectSessionModal.propTypes = {
  courseNumber: PropTypes.string.isRequired,
};

export default SelectSessionModal;
