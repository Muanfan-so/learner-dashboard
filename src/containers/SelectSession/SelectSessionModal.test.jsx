import React from 'react';
import { shallow } from 'enzyme';

import hooks from './hooks';
import SelectSessionModal from './SelectSessionModal';

jest.mock('./hooks', () => ({
  __esModule: true,
  default: jest.fn(),
}));

const hookReturn = {
  entitlementSessions: [],
  showSessionModal: true,
  closeSessionModal: jest.fn().mockName('useSelectSession.closeSessionModal'),
  showLeaveSessionInSessionModal: true,
  courseTitle: 'course-title: unit test save life',
};

const courseNumber = 'my-test-course-number';

const availableSessions = [
  { startDate: '1/2/2000', endDate: '1/2/2020', courseNumber },
  { startDate: '2/3/2000', endDate: '2/3/2020', courseNumber },
  { startDate: '3/4/2000', endDate: '3/4/2020', courseNumber },
];

describe('SelectSessionModal', () => {
  describe('snapshot', () => {
    test('empty modal with leave option ', () => {
      hooks.mockReturnValueOnce({
        ...hookReturn,
      });
      expect(shallow(<SelectSessionModal courseNumber={courseNumber} />)).toMatchSnapshot();
    });

    test('modal with leave option ', () => {
      hooks.mockReturnValueOnce({
        ...hookReturn,
        entitlementSessions: [...availableSessions],
      });
      expect(shallow(<SelectSessionModal courseNumber={courseNumber} />)).toMatchSnapshot();
    });

    test('modal without leave option ', () => {
      hooks.mockReturnValueOnce({
        ...hookReturn,
        entitlementSessions: [...availableSessions],
        showLeaveSessionInSessionModal: false,
      });
      expect(shallow(<SelectSessionModal courseNumber={courseNumber} />)).toMatchSnapshot();
    });
  });
});
