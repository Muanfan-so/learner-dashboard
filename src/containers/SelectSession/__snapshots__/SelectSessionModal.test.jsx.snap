// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectSessionModal snapshot empty modal with leave option  1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    size="md"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <h4>
      Change or leave session?
    </h4>
    <Form.Group>
      <Form.Label>
        When you change to a different session any course progress or grades from your current session will be lost.
      </Form.Label>
      <Form.RadioSet
        name="sessions"
      >
        <Form.Radio
          value="leave"
        >
          Leave session
        </Form.Radio>
      </Form.RadioSet>
    </Form.Group>
    <ActionRow>
      <Button
        onClick={[MockFunction useSelectSession.closeSessionModal]}
        variant="tertiary"
      >
        Nevermind
      </Button>
      <Button>
        Confirm Session
      </Button>
    </ActionRow>
  </div>
</ModalDialog>
`;

exports[`SelectSessionModal snapshot modal with leave option  1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    size="md"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <h4>
      Change or leave session?
    </h4>
    <Form.Group>
      <Form.Label>
        When you change to a different session any course progress or grades from your current session will be lost.
      </Form.Label>
      <Form.RadioSet
        name="sessions"
      >
        <Form.Radio
          key="1/2/2000"
          value="1/2/2000"
        >
          1/2/2000
           - 
          1/2/2020
        </Form.Radio>
        <Form.Radio
          key="2/3/2000"
          value="2/3/2000"
        >
          2/3/2000
           - 
          2/3/2020
        </Form.Radio>
        <Form.Radio
          key="3/4/2000"
          value="3/4/2000"
        >
          3/4/2000
           - 
          3/4/2020
        </Form.Radio>
        <Form.Radio
          value="leave"
        >
          Leave session
        </Form.Radio>
      </Form.RadioSet>
    </Form.Group>
    <ActionRow>
      <Button
        onClick={[MockFunction useSelectSession.closeSessionModal]}
        variant="tertiary"
      >
        Nevermind
      </Button>
      <Button>
        Confirm Session
      </Button>
    </ActionRow>
  </div>
</ModalDialog>
`;

exports[`SelectSessionModal snapshot modal without leave option  1`] = `
<ModalDialog
  hasCloseButton={false}
  isOpen={true}
  onClose={[MockFunction hooks.nullMethod]}
  title=""
>
  <div
    className="bg-white p-3 rounded shadow"
    size="md"
    style={
      Object {
        "textAlign": "start",
      }
    }
  >
    <h4>
      Select a session to access course-title: unit test save life
    </h4>
    <Form.Group>
      <Form.Label>
        Remember, if you change your mind you have 2 weeks to unenroll and reclaim your entitlement.
      </Form.Label>
      <Form.RadioSet
        name="sessions"
      >
        <Form.Radio
          key="1/2/2000"
          value="1/2/2000"
        >
          1/2/2000
           - 
          1/2/2020
        </Form.Radio>
        <Form.Radio
          key="2/3/2000"
          value="2/3/2000"
        >
          2/3/2000
           - 
          2/3/2020
        </Form.Radio>
        <Form.Radio
          key="3/4/2000"
          value="3/4/2000"
        >
          3/4/2000
           - 
          3/4/2020
        </Form.Radio>
      </Form.RadioSet>
    </Form.Group>
    <ActionRow>
      <Button
        onClick={[MockFunction useSelectSession.closeSessionModal]}
        variant="tertiary"
      >
        Nevermind
      </Button>
      <Button>
        Confirm Session
      </Button>
    </ActionRow>
  </div>
</ModalDialog>
`;
